Src/WindowsServer/WindowsServer.Shared/AppServicesHeartbeatTelemetryModule.cs

```diff
@@ -0,0 +1,150 @@
+            {
+                lock (this.lockObject)
+                {
+                    if (!this.isInitialized)
+                    {
+                        this.AddAppServiceEnvironmentVariablesToHeartbeat(this.GetHeartbeatPropertyManager(), false);
+
+                        this.isInitialized = true;
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Ensure we've cleaned up our static Instance.
+        /// </summary>
+        public void Dispose()
+        {
+            // ensure our Instance variable is not kept around.
+            AppServicesHeartbeatTelemetryModule.instance = null;
+        }
+
+        /// <summary>
+        /// Signal the AppServicesHeartbeatTelemetryModule to update the values of the 
+        /// Environment variables we use in our heartbeat payload.
+        /// </summary>
+        internal void UpdateHeartbeatWithAppServiceEnvVarValues()
+        {
+            if (this.isInitialized)
+            {
+                this.AddAppServiceEnvironmentVariablesToHeartbeat(this.GetHeartbeatPropertyManager(), isUpdateOperation: true);
+            }
+        }
+
+        private IHeartbeatPropertyManager GetHeartbeatPropertyManager()
+        {
+            IHeartbeatPropertyManager hbeatManager = null;
+            var telemetryModules = TelemetryModules.Instance;
+
+            try
+            {
+                foreach (var module in telemetryModules.Modules)
+                {
+                    if (module is IHeartbeatPropertyManager hman)
+                    {
+                        hbeatManager = hman;
+                    }
+                }
+            }
+            catch (Exception hearbeatManagerAccessException)
+            {
+                WindowsServerEventSource.Log.HeartbeatManagerAccessFailure(hearbeatManagerAccessException.Message);
+            }
+
+            return hbeatManager;
+        }
+
+        private void AddAppServiceEnvironmentVariablesToHeartbeat(IHeartbeatPropertyManager hbeatManager, bool isUpdateOperation = false)
+        {
+            var telemetryModules = TelemetryModules.Instance;
+            Dictionary<string, string> hbeatProps = new Dictionary<string, string>();
+
+            foreach (var kvp in WebHeartbeatPropertyNameEnvVarMap)
+            {
+                try
+                {
+                    // get the variable, then expand it (otherwise we get the name we queried for in the value)
+                    string hbeatValue = Environment.GetEnvironmentVariable(kvp.Value);
+                    if (!string.IsNullOrEmpty(hbeatValue))
+                    {
+                        hbeatValue = Environment.ExpandEnvironmentVariables(hbeatValue);
+                        string hbeatKey = kvp.Key.ToString();
+                        if (isUpdateOperation)
+                        {
+                            hbeatManager.SetHeartbeatProperty(hbeatKey, hbeatValue);
```

Comment:
Would you expect `hbeatManager` to be `null` ever?

Reply:
Yes, I've added an early exit to counter this possibility.